### ISSUE TITLE
Phrase search does not return the correct highlight

### ISSUE THREAD

**Issue by tao** [meilisearch/meilisearch#1912]
**Describe the bug**
When performing a phrase search, "the seeker seeks the one" as one example on my data. The phrase search will return the first highlight with the word "the" in it... instead of returning the highlight with the entire phrase.

![Screenshot 2021-11-18 at 12 59 29](https://user-images.githubusercontent.com/1446331/142403120-9a76b44f-8168-4d06-9161-2581136b115d.png)

If you take a look at the text itself, you can see the phrase "the seeker seeks the one" exists in the third paragraph but the highlight in the search hits returns the first sentence.

![Screenshot 2021-11-18 at 13 01 28](https://user-images.githubusercontent.com/1446331/142403412-d730e1ba-23fb-4e4e-b18c-ecb7e8230f40.png)

**Expected behavior**
I would expect the phrase search to only return results which includes the exact phrase and only highlight the exact phrase.

**MeiliSearch version:** [e.g. v0.20.0]
```json
{
    "commitSha":"2e2eeb0a424408e27f4223e58afec303c53e1419",
    "commitDate":"2021-10-12T14:47:06+00:00",
    "pkgVersion":"0.23.1"
}
```

**Additional context**
I understand stop words might be used to exclude "the" and "of" from the search, which might help the result... but in our case we want to search dialogue exactly for "the seeker seeks the one" for example, and not "seeker of the one seeks" or another combination of the phrase, for example.

**Comment by tao** [comment]
In my Javascript search config, I am cropping the content by 140 characters because we do have larger chunks of text with many paragraphs, it could be that MeiliSearch finds the correct phrase/sentence but starts cropping at 0-140 instead of cropping and highlighting around the phrase.

**Comment by curquiza** [comment]
Hello @tao 

Thanks for the report, this is really appreciated!

The core team will do their best to investigate and fix this in the next release, in December. Sorry for the inconvenience!

**Comment by gmourier** [comment]
Hello @tao!

It looks like the highlighter is separating all the words in the search phrase, considering them as separated terms and highlighting them one by one. This behavior is interesting when we are out of context of an exact phrase query however what we could consider to solve this problem is to take the whole phrase search query as one "word" and apply it where it matches.

e.g. `"The seeker seeks the One"`

Currently, this is what happens.

```json
{
    "hits": [
        {
            "_formatted": {
                "searchable_field": "This is a long text. <em>The</em> <em>seeker</em> <em>seeks</em> <em>the</em <em>one</em> exact phrase search. <em>The</em> car is red."
            },
            "_matchesInfo": {
                "searchable_field": [
                    {
                        "start": 22,
                        "length": 6
                    },
                    {
                        "start": 43,
                        "length": 2
                    },
                    {
                        "start": 46,
                        "length": 3
                    },
                    {
                        "start": 50,
                        "length": 3
                    },
                    {
                        "start": 54,
                        "length": 5
                    },
                    {
                        "start": 101,
                        "length": 3
                    }
                ]
            }
        }
    ],

}
```

> Note, that _matchesInfo is following the same rule.

It could be something like that when we encounter an exact phrase search within the `q` parameter.


```json
{
    "hits": [
        {
            "_formatted": {
                "searchable_field": "This is a long text. <em>The seeker seeks the one</em> exact phrase search. The car is red."
            },
            "_matchesInfo": {
                "searchable_field": [
                    {
                        "start": 22,
                        "length": 24
                    }
                ]
            }
        }
    ],

}
```

While searching for an exact phrase search and a term. E.g. `"The seeker seeks the one" the` could be

```json
{
    "hits": [
        {
            "_formatted": {
                "searchable_field": "This is a long text. <em>The seeker seeks the one</em> exact phrase search. <em>The</em> car is red."
            },
            "_matchesInfo": {
                "searchable_field": [
                    {
                        "start": 22,
                        "length": 24
                    },
                    ...  //_matchesInfo for `The` query term
                ]
            }
        }
    ],

}
```

I don't know if it is feasible very easily in the current state and maybe there are limits that I don't see.

**Comment by tao** [comment]
Yes, from our work so far it seems like the default search works as a "match any word", while phrase search currently works a lot more like "match all words"... which is something we do want to include as some optional functionality for the user. 

However, we also really need a "match exact phrase" search, that would only return results that include (and highlight) the exact phrase. Even if we have to highlight it ourselves that's fine... So it wouldn't highlight loose words like `The` or `the One` in that case.

I did also reach out to the support chat recently about this "all words", "any word", "exact phrase" but haven't heard back yet... but I noticed in some other open issues that phrase search is being worked on and maybe it's related?

**Comment by gmourier** [comment]
Hi @tao! 👋

The phrase search pattern allows two things:

- "The car is red" -> The returned documents will contain EXACTLY the terms in that order. 
- "The" "car" "is" "red". -> The documents returned will contain the terms in no particular order.

We have a product discussion concerning the [policy for matching words](https://github.com/meilisearch/product/discussions/255) (as I may know it does not directly concern the phrase search but "simple" query terms behavior).

**Comment by ManyTheFish** [comment]
Hello! Sorry, I took a bit of time to respond 😖.

I honestly think that this is not a bug. The current highlighter of meilisearch is "dumb". It is not possible to find the best match of the query in a field. It basically says: "this word is a match".
However, it's possible to improve it, forcing it to find the best match in each field instead of the first one. We could have some performances drawbacks, but it would make the crop way more relevant.

I think this issue can be fixed relatively easily and could be a `good first issue`. 🤝

Thanks @tao, for your report! 👍

**Comment by tao** [comment]
Thanks, I understand... I think it's maybe a bit more of an inconsistency in how I would expect phrase search to work, or maybe the highlighting on phrase search could be stricter, but since it's not that relevant I did discuss that more in the policy on matching words discussion.

It's more of a cropping issue than a highlighting issue? So basically if you include any common word (the, of, one, to) the highlighter would start cropping from the first instance of the word instead of cropping around the best match. 

Maybe it could be improved then by letting the cropping function ignore stop words?

**Comment by ManyTheFish** [comment]
Yes, you're right, it's a cropping issue, but the cropping function uses the highlighter to know where the matches are in an attribute. For now, it stops on the first match found and not on the more relevant match. What we could do is crop around the biggest "match density" in a field. 🤔

**Comment by gmourier** [comment]
Postponed to a future version due to lack of time on your side to treat it well.

**Comment by curquiza** [comment]
Hello everyone, 

Checked with @ManyTheFish, this issue contains two problems
- a bug, that has been fixed in v0.25.0
- a feature request -> crop by density instead of cropping around the first match. Since it's a design change and not a bug, this is handled on the product repository. See this discussion regarding this: https://github.com/meilisearch/product/discussions/350

@ManyTheFish has also started some discoveries and tests in https://github.com/meilisearch/meilisearch/pull/2130

I'm closing this issue then since the discussion is now on the product's side :) Feel free to give your feedback on it!
